ext/storage_sources/s3_store/s3_connection.cpp

```diff
@@ -99,6 +101,35 @@ S3Connection::DeleteObject(const std::string &bucketName, const std::string &obj
     }
 }
 
+/*
+ * ObjectExists --
+ *     Checks whether an object with the given key exists in the S3 bucket.
+ */
+int
+S3Connection::ObjectExists(
+  const std::string &bucketName, const std::string &objectKey, bool &exists) const
+{
+    exists = false;
+
+    Aws::S3Crt::Model::HeadObjectRequest request;
+    request.SetBucket(bucketName);
+    request.SetKey(objectKey);
+    Aws::S3Crt::Model::HeadObjectOutcome outcome = m_S3CrtClient.HeadObject(request);
+
+    /*
+     * If an object with the given key does not exist the HEAD request will return a 404.
+     * https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html Do not fail in this case.
+     */
+    if (outcome.IsSuccess()) {
+        exists = true;
+        return (0);
+    } else if (outcome.GetError().GetResponseCode() == Aws::Http::HttpResponseCode::NOT_FOUND)
+        return (0);
+    else
+        /* There are other reasons ObjectExists might fail. This can be changed in the future. */
+        return (static_cast<int>(outcome.GetError().GetResponseCode()));
+}
```

Comment:
If we reach at the end of the function we still want to return a failure code. Maybe -1 for now.

Reply:
Do you mean after the if-statement or in that `else` clause instead of the HTTP response code?

Reply:
Do you mean -1 instead of the HTTP code?